### PR TITLE
fix: delay emitting NotifyIcon events on Windows

### PR DESCRIPTION
**Pull request by erickzhao**
#### Description of Change
Following prior work done in #25836, this PR delays emitting Balloon events from the `Tray` module to avoid entering V8 during a WndProc callback.

##### Implementation Details

Since `TrayIcon` objects are deleted when the `Tray` gets garbage collected, we don't want the callback to be processed if the icon is gone. This PR uses `WeakPtr` references to achieve this end.

This also means that the affected events won't be emitted if the WndProc callback is entered, but the `TrayIcon` is destroyed. This behaviour seems in line with Electron's existing `Tray` API design, as a reference to the `Tray` needs to be maintained for its API to be used at all.

cc @nornagon 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Fixed a rare crash on Windows that could occur when emitting certain `Tray` events.

### CODE REVIEW

**Review by codebytere — APPROVED** [review]

File: shell/browser/ui/win/notify_icon_host.h [hunk 1]
```diff
@@ -10,6 +10,7 @@
 #include <vector>
 
 #include "base/macros.h"
+#include "base/memory/weak_ptr.h"
```
Comment (nornagon) [inline shell/browser/ui/win/notify_icon_host.h:13]: This `#include` should go in the cc file if it's not needed in the header

File: shell/browser/ui/win/notify_icon.h [hunk 2]
```diff
@@ -101,6 +103,8 @@ class NotifyIcon : public TrayIcon {
   // Context menu associated with this icon (if any).
   std::unique_ptr<views::MenuRunner> menu_runner_;
 
+  base::WeakPtrFactory<NotifyIcon> weak_factory_;
```
Comment (nornagon) [inline shell/browser/ui/win/notify_icon.h:106]:
```suggestion
  base::WeakPtrFactory<NotifyIcon> weak_factory_{this};
```

File: shell/browser/ui/win/notify_icon.cc [hunk 3]
```diff
@@ -49,7 +49,11 @@ NotifyIcon::NotifyIcon(NotifyIconHost* host,
                        HWND window,
                        UINT message,
                        GUID guid)
-    : host_(host), icon_id_(id), window_(window), message_id_(message) {
+    : host_(host),
+      icon_id_(id),
+      window_(window),
+      message_id_(message),
+      weak_factory_(this) {
```
Comment (nornagon) [inline shell/browser/ui/win/notify_icon.cc:56]: The modern style for initializing the weak factory is to use a default-initializer (see below)

```suggestion
      message_id_(message) {
```

File: shell/browser/ui/win/notify_icon_host.cc [hunk 4]
```diff
@@ -142,14 +143,14 @@ LRESULT CALLBACK NotifyIconHost::WndProc(HWND hwnd,
...
-    NotifyIcon* win_icon = NULL;
+    base::WeakPtr<NotifyIcon> win_icon = NULL;
 
     // Find the selected status icon.
     for (NotifyIcons::const_iterator i(notify_icons_.begin());
          i != notify_icons_.end(); ++i) {
       auto* current_win_icon = static_cast<NotifyIcon*>(*i);
       if (current_win_icon->icon_id() == wparam) {
-        win_icon = current_win_icon;
+        win_icon = current_win_icon->GetWeakPtr();
```
Comment (nornagon) [inline shell/browser/ui/win/notify_icon_host.cc:153]: keep this loop as it was before and call `->GetWeakPtr()` after finding the icon. That way we call `GetWeakPtr` only once instead of _n_ times.

File: shell/browser/ui/win/notify_icon_host.cc [hunk 5]
```diff
@@ -162,15 +163,30 @@ LRESULT CALLBACK NotifyIconHost::WndProc(HWND hwnd,
 
     switch (lparam) {
       case NIN_BALLOONSHOW:
-        win_icon->NotifyBalloonShow();
+        content::GetUIThreadTaskRunner({})->PostTask(
+            FROM_HERE, base::BindOnce(
+                           [](base::WeakPtr<NotifyIcon> wicon) {
+                             wicon->NotifyBalloonShow();
+                           },
+                           win_icon));
```
Comment (nornagon) [inline shell/browser/ui/win/notify_icon_host.cc:171]: The code as is will fail because it doesn't check `if (wicon)`. I recommend using the member-function variant of `base::Bind` instead, which handles this automatically:

```suggestion
        content::GetUIThreadTaskRunner({})->PostTask(
            FROM_HERE, base::BindOnce(
                           &NotifyIcon::NotifyBalloonShow,
                           win_icon));
```

(see [base/callback.h:46-49](https://source.chromium.org/chromium/chromium/src/+/master:base/callback.h;l=46-49;drc=788a3897a44bde5263e83172c84ebb85d99b5080))

Same for the cases below.

**Review by nornagon — CHANGES_REQUESTED** [review]

File: shell/browser/ui/win/notify_icon_host.cc [hunk 6]
```diff
@@ -160,17 +162,35 @@ LRESULT CALLBACK NotifyIconHost::WndProc(HWND hwnd,
...
+    // that the callback is aware if the NotifyIcon gets
+    // garbage-collected. This occurs when the tray gets
+    // GC'd, and the BALLOON events below will not emit.
+    base::WeakPtr<NotifyIcon> win_icon_weak = win_icon->GetWeakPtr();
+
+    int modifiers = GetKeyboardModifers();
+    bool left_button_click =
+        (lparam == WM_LBUTTONDOWN || lparam == WM_LBUTTONDBLCLK);
+    bool double_button_click =
+        (lparam == WM_LBUTTONDBLCLK || lparam == WM_RBUTTONDBLCLK);
```
Comment (nornagon) [inline shell/browser/ui/win/notify_icon_host.cc:175]: these can stay inlined

File: shell/browser/ui/win/notify_icon_host.cc [hunk 7]
```diff
@@ -169,10 +169,6 @@ LRESULT CALLBACK NotifyIconHost::WndProc(HWND hwnd,
     base::WeakPtr<NotifyIcon> win_icon_weak = win_icon->GetWeakPtr();
 
     int modifiers = GetKeyboardModifers();
```
Comment (nornagon) [inline shell/browser/ui/win/notify_icon_host.cc:171]: this too i think (sorry!)

File: shell/browser/ui/win/notify_icon_host.cc [hunk 8]
```diff
@@ -34,7 +36,7 @@ bool IsWinPressed() {
          ((::GetKeyState(VK_RWIN) & 0x8000) == 0x8000);
 }
 
-int GetKeyboardModifers() {
+int GetKeyboardModifiers() {
```
Comment (erickzhao) [inline shell/browser/ui/win/notify_icon_host.cc:39]: Unrelated typo fix in my last commit.

**Review by nornagon — APPROVED** [review]

**Review by zcbenz — APPROVED** [review]
